search/searcher/search_nested.go

```diff
@@ -0,0 +1,154 @@
+//  Copyright (c) 2019 Couchbase, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 		http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package searcher
+
+import (
+	"github.com/blevesearch/bleve/index"
+	"github.com/blevesearch/bleve/search"
+)
+
+type depthFilterContext struct {
+	termAPMap  map[string][]search.ArrayPositions
+	apLists    [][]search.ArrayPositions
+	arrayDepth int
+}
+
+func NewNestedArraySearcher(indexReader index.IndexReader, depth int,
+	qsearchers []search.Searcher, options search.SearcherOptions) (
+	search.Searcher, error) {
+	conjunctionSearcher, err := NewConjunctionSearcher(indexReader, qsearchers,
+		options)
+	if err != nil {
+		return nil, err
+	}
+
+	rv := depthFilterContext{arrayDepth: depth}
+
+	return NewFilteringSearcher(conjunctionSearcher,
+		rv.buildArrayDepthFilter()), nil
+}
+
+// buildArrayDepthFilter returns a filter that checks whether
+// all the search terms are sharing any common array position in
+// the document match.
+func (c *depthFilterContext) buildArrayDepthFilter() FilterFunc {
+	return c.checkArrayDepth
+}
+
+func (c *depthFilterContext) checkArrayDepth(d *search.DocumentMatch) bool {
+	if d == nil || len(d.FieldTermLocations) == 0 {
+		return false
+	}
+
+	// reset and resuse
+	if len(c.termAPMap) == 0 {
+		c.termAPMap = make(map[string][]search.ArrayPositions, len(d.FieldTermLocations))
+	} else {
+		for k := range c.termAPMap {
+			list := c.termAPMap[k]
+			c.termAPMap[k] = list[:0]
+		}
+	}
+
+	termList := make(map[string]struct{}, 1)
+	for _, ftl := range d.FieldTermLocations {
+		if len(ftl.Location.ArrayPositions) == c.arrayDepth {
+			c.termAPMap[ftl.Term] = append(c.termAPMap[ftl.Term],
+				ftl.Location.ArrayPositions)
+		}
+		termList[ftl.Term] = struct{}{}
+	}
+
+	// exit early if any of the search terms are not found
+	// at the given array depth.
+	for term := range termList {
+		if _, ok := c.termAPMap[term]; !ok {
+			return false
```

Comment:
Again, relating to my previous comment, I expected matches could be deeper than the requested depth, and that still makes sense.  Is there a use case for giving the user the choice of either "exact depth" vs "at least depth"?

Reply:
hm. the intent was only consider the matches at the asked depth/exact depth and then check their array positions equality. 
Need to think more on the "exact depth" vs "at least depth" use cases.

at least depth - gives an impression that user isn't aware of the document structure or doesn't care about the match positions (except the exclusion factor of atleast) - but then he can stick to the blind search with out the depth parameter ?